felix/fv/policy_grouping_test.go

```diff
@@ -0,0 +1,196 @@
+// Copyright (c) 2023 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build fvtests
+
+package fv_test
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	v3 "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
+
+	"github.com/projectcalico/calico/felix/fv/connectivity"
+	"github.com/projectcalico/calico/felix/fv/infrastructure"
+	"github.com/projectcalico/calico/felix/fv/workload"
+	"github.com/projectcalico/calico/felix/rules"
+	"github.com/projectcalico/calico/libcalico-go/lib/apiconfig"
+	client "github.com/projectcalico/calico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/calico/libcalico-go/lib/options"
+)
+
+var _ = infrastructure.DatastoreDescribe("policy grouping tests", []apiconfig.DatastoreType{apiconfig.Kubernetes, apiconfig.EtcdV3}, func(getInfra infrastructure.InfraFactory) {
+	var (
+		tc     infrastructure.TopologyContainers
+		infra  infrastructure.DatastoreInfra
+		client client.Interface
+		w      [3]*workload.Workload
+		cc     *connectivity.Checker
+	)
+
+	BeforeEach(func() {
+		infra = getInfra()
+		opts := infrastructure.DefaultTopologyOptions()
+		tc, client = infrastructure.StartNNodeTopology(1, opts, infra)
+		for i := range w {
+			w[i] = workload.Run(
+				tc.Felixes[0],
+				fmt.Sprintf("w%d", i),
+				"default",
+				fmt.Sprintf("10.65.0.%d", i+2),
+				"8080",
+				"tcp",
+			)
+			w[i].ConfigureInInfra(infra)
+		}
+		cc = &connectivity.Checker{}
+
+		// Add a default-allow policy so that we know that our deny rule is
+		// effective.
+		pol := v3.NewGlobalNetworkPolicy()
+		pol.Name = "pol-default-allow"
+		// Need a selector that matches all endpoints but is different to the
+		// ones used in the test.
+		pol.Spec.Selector = "!has(unknown)"
+		pol.Spec.Order = floatPtr(200)
+		pol.Spec.Ingress = []v3.Rule{
+			{
+				Action: v3.Allow,
+			},
+		}
+		pol.Spec.Egress = []v3.Rule{
+			{
+				Action: v3.Allow,
+			},
+		}
+		pol, err := client.GlobalNetworkPolicies().Create(context.TODO(), pol, options.SetOptions{})
```

Comment:
Any pre-existing, top-level context that could be used instead of `TODO` that would respond to interrupts?

Reply:
Yeah, should really use a timeout; annoying that the test harness doesn't provide one (AFAICT)